airflow/jobs/scheduler_job_runner.py

```diff
@@ -1548,33 +1555,51 @@ def _update_state(dag: DAG, dag_run: DagRun):
 
         span = Trace.get_current_span()
         for dag_run in dag_runs:
-            dag = dag_run.dag = cached_get_dag(dag_run.dag_id)
-
+            dag_id = dag_run.dag_id
+            run_id = dag_run.run_id
+            backfill_id = dag_run.backfill_id
+            backfill = dag_run.backfill
+            dag = dag_run.dag = cached_get_dag(dag_id)
             if not dag:
                 self.log.error("DAG '%s' not found in serialized_dag table", dag_run.dag_id)
                 continue
-            active_runs = active_runs_of_dags[dag_run.dag_id]
-
-            if dag.max_active_runs and active_runs >= dag.max_active_runs:
-                self.log.debug(
-                    "DAG %s already has %d active runs, not moving any more runs to RUNNING state %s",
-                    dag.dag_id,
-                    active_runs,
-                    dag_run.execution_date,
-                )
-            else:
-                if span.is_recording():
-                    span.add_event(
-                        name="dag_run",
-                        attributes={
-                            "run_id": dag_run.run_id,
-                            "dag_id": dag_run.dag_id,
-                            "conf": str(dag_run.conf),
-                        },
+            active_runs = active_runs_of_dags[(dag_id, backfill_id)]
+            if backfill_id is not None:
+                if active_runs >= backfill.max_active_runs:
```

Comment:
This means in practice we have a totally separate max_active_runs for scheduled and each backfill?

Intuitively, it feels like the DAGs max_active_runs should be honored across types.

Reply:
yeah, tricky to decide the right interface. the idea with doing it separately is you have complete control over both kinds of runs.

if they are mixed in together, the number of concurrent backfill dag runs depends on how many prod runs are running.  backfill runs could get starved out by prod runs (at least in theory) -- not likely i'd bet.